src/analysis/constant_speed.lean

```diff
@@ -0,0 +1,281 @@
+  ∀ ⦃x⦄ (hx : x ∈ s) ⦃y⦄ (hy : y ∈ s), variation_on_from_to f s x y = l * (y - x)) :=
+begin
+  split,
+  { rintro h, refine ⟨h.has_locally_bounded_variation_on, λ x xs y ys, _⟩,
+    rw has_constant_speed_on_with_iff_ordered at h,
+    rcases le_total x y with xy|yx,
+    { rw [variation_on_from_to.eq_of_le f s xy, h xs ys xy,
+          ennreal.to_real_of_real (mul_nonneg l.prop (sub_nonneg.mpr xy))], },
+    { rw [variation_on_from_to.eq_of_ge f s yx, h ys xs yx ,
+          ennreal.to_real_of_real  (mul_nonneg l.prop (sub_nonneg.mpr yx)),
+          mul_comm ↑l, mul_comm ↑l, ←neg_mul, neg_sub], }, },
+  { rw has_constant_speed_on_with_iff_ordered,
+    rintro h x xs y ys xy,
+    rw [←h.2 xs ys, variation_on_from_to.eq_of_le f s xy,
+        ennreal.of_real_to_real (h.1 x y xs ys)], },
+end
+
+lemma has_constant_speed_on_with.union {t : set ℝ}
+  (hfs : has_constant_speed_on_with f s l) (hft : has_constant_speed_on_with f t l)
+  {x : ℝ} (hs : is_greatest s x) (ht : is_least t x) : has_constant_speed_on_with f (s ∪ t) l :=
+begin
+  rw has_constant_speed_on_with_iff_ordered at hfs hft ⊢,
+  rintro z (zs|zt) y (ys|yt) zy,
+  { have : (s ∪ t) ∩ Icc z y = (s ∩ Icc z y), by
+    { ext w, split,
+      { rintro ⟨(ws|wt), zw, wy⟩,
+        { exact ⟨ws, zw, wy⟩, },
+        { exact ⟨(le_antisymm (wy.trans (hs.2 ys)) (ht.2 wt)).symm ▸ hs.1, zw, wy⟩, }, },
+      { rintro ⟨ws,zwy⟩, exact ⟨or.inl ws, zwy⟩, }, },
+    rw [this, hfs zs ys zy], },
+  { have : (s ∪ t) ∩ Icc z y = (s ∩ Icc z x) ∪ (t ∩ Icc x y), by
+    { ext w, split,
+      { rintro ⟨(ws|wt), zw, wy⟩,
+        exacts [or.inl ⟨ws, zw, hs.2 ws⟩, or.inr ⟨wt, ht.2 wt, wy⟩], },
+      { rintro (⟨ws, zw, wx⟩|⟨wt, xw, wy⟩),
+        exacts [⟨or.inl ws, zw, wx.trans (ht.2 yt)⟩, ⟨or.inr wt, (hs.2 zs).trans xw, wy⟩], }, },
+    rw [this,
+        @evariation_on.union _ _ _ _ f _ _ x ,
+        hfs zs hs.1 (hs.2 zs), hft ht.1 yt (ht.2 yt),
+        ←ennreal.of_real_add (mul_nonneg l.prop (sub_nonneg.mpr (hs.2 zs)))
+                             (mul_nonneg l.prop (sub_nonneg.mpr (ht.2 yt))) ],
+    ring_nf,
+    exacts [⟨⟨hs.1, hs.2 zs, le_rfl⟩, λ w ⟨ws, zw, wx⟩, wx⟩,
+            ⟨⟨ht.1, le_rfl, ht.2 yt⟩, λ w ⟨wt, xw, wy⟩, xw⟩], },
+  { cases le_antisymm zy ((hs.2 ys).trans (ht.2 zt)),
+    simp only [Icc_self, sub_self, mul_zero, ennreal.of_real_zero],
+    exact evariation_on.subsingleton _ (λ _ ⟨_, uz⟩ _ ⟨_, vz⟩, uz.trans vz.symm), },
+  { have : (s ∪ t) ∩ Icc z y = (t ∩ Icc z y), by
+    { ext w, split,
+      { rintro ⟨(ws|wt), zw, wy⟩,
+        { exact ⟨(le_antisymm ((ht.2 zt).trans zw) (hs.2 ws)) ▸ ht.1, zw, wy⟩, },
+        { exact ⟨wt, zw, wy⟩, }, },
+      { rintro ⟨wt, zwy⟩, exact ⟨or.inr wt, zwy⟩, }, },
+    rw [this, hft zt yt zy], }
+end
+
+lemma has_constant_speed_on_with.Icc_Icc {x y z : ℝ}
+  (hfs : has_constant_speed_on_with f (Icc x y) l) (hft : has_constant_speed_on_with f (Icc y z) l)
+  (xy : x ≤ y) (yz : y ≤ z) : has_constant_speed_on_with f (Icc x z) l :=
```

Comment:
you shouldn't need `xy` and `yz` here, I guess.

Reply:
Good catch! working on it